Fully deprecate StopCh

mjudeikis opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind cleanup

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:



#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fully remove PostStartHookContext.StopCh
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

mjudeikis committed:
fully remove StopCh

Signed-off-by: Mangirdas Judeikis <mangirdas@judeikis.lt>

mjudeikis commented:
Should be last of StopCh...

sttts commented on staging/src/k8s.io/apiserver/pkg/server/hooks.go:
```diff
@@ -49,11 +49,6 @@ type PreShutdownHookFunc func() error
-	//
-	// Deprecated: use the PostStartHookContext itself instead, it contains a context that
-	// gets cancelled when the server stops. StopCh keeps getting provided for existing code.
-	StopCh <-chan struct{}
```
when did we deprecate it? Longer ago?

sttts replied:
> Fully deprecate PostStartHookContext.StopCh

Fully remove it, no?

mjudeikis replied:
Its deprecated ages go. I found this message and just took on myself to do the cleaning :)

mjudeikis replied:
updated release notes

mjudeikis replied:
updated git commit message too

aojea commented:
linter complains in
```
ERROR: staging/src/k8s.io/apiserver/pkg/server/config.go:908:24: Error return value of `c.FlowControl.Run` is not checked (errcheck)
ERROR: 			go c.FlowControl.Run(hookContext.Done())
ERROR: 			                    ^
```

mjudeikis commented:
> go c.FlowControl.Run(hookContext.Done())

I was bit cautions here as we do this all around the place and hints shows this, but does not block.
I changed it now, but this changes a behaviour a bit, where, if previously it fail silently and moves on. If we check and return - it will fail server startup? 

And I was not sure if author intended this to be like this or now. It looked strange why inside `AddPostStartHook` (which is go routine on its own) we kick-off new go routine :shrug: 

Either way, changed now. Lets see what tests shows.

mjudeikis commented:
wow, this just exploded in CI. Lets ee if this real or just sunday rush hour.
/retest

aojea commented:
> wow, this just exploded in CI. Lets ee if this real or just sunday rush hour.
> /retest

please click on the jobs before retesting , all jobs testing failing means something is wrong.

It seems this PR is breaking the apiserver start, apiserver never gets pass the bootsrap phase https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/127341/pull-kubernetes-e2e-kind/1837906716155449344/artifacts/kind-control-plane/containers/kube-apiserver-kind-control-plane_kube-system_kube-apiserver-2179625b5987564cc3d2d04bb84be2ce671238b70a28bc5fb1ecbb7c03cd167c.log

> 2024-09-22T17:43:40.207515019Z stderr F I0922 17:43:40.207443       1 httplog.go:134] "HTTP" verb="GET" URI="/readyz" latency="2.360446ms" userAgent="kube-apiserver/v1.32.0 (linux/amd64) kubernetes/747066c" audit-ID="1fa536a2-42e2-4e9c-b34f-95af099752d0" srcIP="[::1]:59452" apf_pl="exempt" apf_fs="exempt" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="2.039863ms" resp=500
2024-09-22T17:43:40.305483732Z stderr F I0922 17:43:40.305317       1 healthz.go:278] poststarthook/priority-and-fairness-config-consumer check failed: readyz
2024-09-22T17:43:40.30553272Z stderr F [-]poststarthook/priority-and-fairness-config-consumer failed: not finished

Jefftree commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -904,9 +904,8 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
-			go c.FlowControl.Run(context.StopCh)
-			return nil
+		err := s.AddPostStartHook(priorityAndFairnessConfigConsumerHookName, func(hookContext PostStartHookContext) error {
+			return c.FlowControl.Run(hookContext.Done())
```
this is a goroutine. We definitely should not be returning here

mjudeikis replied:
reverted back to original change

Jefftree commented:
> linter complains in
> 
> ```
> ERROR: staging/src/k8s.io/apiserver/pkg/server/config.go:908:24: Error return value of `c.FlowControl.Run` is not checked (errcheck)
> ERROR: 			go c.FlowControl.Run(hookContext.Done())
> ERROR: 			                    ^
> ```

It's an optional linter and imo we're not making the code base worse by deprecating stopch? I think properly fixing the lint will involve changing the `FlowControl.Run()` function and interface to not return an error.

ref: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go#L371-L391

mjudeikis commented:
> It's an optional linter and imo we're not making the code base worse by deprecating stopch? I think properly fixing the lint will involve changing the `FlowControl.Run()` function and interface to not return an error.

I was thinking same just got persuaded :D reverted now to the original change. FlowControl indeed need fixing and might need to be a separate PR.

mjudeikis commented:
> please click on the jobs before retesting , all jobs testing failing means something is wrong.

As I suspected returning there will explode things. As this is a hint, once merged will create an issue to fix FlowControl as suggested by @Jefftree . One of "good first issue" ones.

sttts commented:
/lgtm
/approve
